codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -508,12 +513,14 @@ class JsonParserGenerator(
                         "Builder" to symbolProvider.symbolForBuilder(shape),
                     )
                     deserializeStructInner(shape.members())
-                    // Only call `build()` if the builder is not fallible. Otherwise, return the builder.
-                    if (returnSymbolToParse.isUnconstrained) {
-                        rust("Ok(Some(builder))")
-                    } else {
-                        rust("Ok(Some(builder.build()))")
+                    val builder = builderInstantiator.finalizeBuilder(
+                        "builder", shape,
+                    ) {
+                        rustTemplate(
+                            """|err|#{Error}::custom_source("Response was invalid", err)""", *codegenScope,
```

Comment:
This doesn't look correct to me. This only applies to clients, not servers. It only works because `ServerBuilderInstantiator` does not rely on the passed in `mapErr` writable.

Reply:
mapErr defines how to convert the error back into the caller error—since the server never returns an error, it doesn't need to use it.